openghg/retrieve/icos/_retrieve.py

```diff
@@ -0,0 +1,223 @@
+        dataframe = dataframe.dropna(axis="index")
+
+        if not dataframe.index.is_monotonic_increasing:
+            dataframe = dataframe.sort_index()
+
+        spec = metadata["species"]
+        rename_cols = {
+            "stdev": spec + " variability",
+            "nbpoints": spec + " number_of_observations",
+        }
+
+        dataframe = dataframe.rename(columns=rename_cols).set_index("timestamp")
+        dataframe.index.name = "time"
+
+        dataset = dataframe.to_xarray()
+        dataset.attrs.update(metadata)
+
+        # TODO - do we need both attributes and metadata here?
+        standardised_data[spec] = {
+            "metadata": metadata,
+            "data": dataset,
+            "attributes": metadata,
+        }
+
+    # Make sure everything is CF compliant
+    standardised_data = assign_attributes(data=standardised_data)
+
+    return standardised_data
+
+
+def _extract_metadata(meta: List, site_metadata: Dict) -> Dict:
+    """Extract metadata from the list of pandas DataFrames that are
+    returned by the ICOS-CP pylib Dobj method.
+
+    Args:
+        meta: List of dataframes containing metadata
+        site_metadata: Dictionary of site metadata
+    Returns:
+        dict: Dictionary of metadata
+    """
+    # From the ICOS documentation
+    # https://icos-carbon-portal.github.io/pylib/modules/#dobjinfo
+    # info[0] -> information about the dobj like, url, specification, number of rows, related file name.
+    # info[1] -> information about the data like colName, value type, unit, kind
+    # info[2] -> information about the station, where the data was obtained. Name, id, lat, lon etc..
+    spec_data = meta[0]
+    measurement_data = meta[1]
+    site_data = meta[2]
+
+    metadata = {}
+
+    metadata["dobj_pid"] = _get_value(df=spec_data, col="dobj", index=0)
+    metadata["species"] = _get_value(df=measurement_data, col="colName", index=4)
+
+    metadata["meas_type"] = _get_value(df=measurement_data, col="valueType", index=4)
+    metadata["units"] = _get_value(df=measurement_data, col="unit", index=4)
+
+    site = _get_value(df=site_data, col="stationId", index=0)
+
+    metadata["site"] = site
+    metadata["station_long_name"] = _get_value(df=site_data, col="stationName", index=0)
+    metadata["sampling_height"] = _get_value(df=site_data, col="samplingHeight", index=0)
+    metadata["inlet"] = _get_value(df=site_data, col="samplingHeight", index=0)
+    metadata["station_latitude"] = _get_value(df=site_data, col="latitude", index=0)
+    metadata["station_longitude"] = _get_value(df=site_data, col="longitude", index=0)
+    metadata["elevation"] = _get_value(df=site_data, col="elevation", index=0)
+
+    site_specific = site_metadata[site.upper()]
+    metadata["data_owner"] = f"{site_specific['firstName']} {site_specific['lastName']}"
+    metadata["data_owner_email"] = site_specific["email"]
+    metadata["station_height_masl"] = site_specific["eas"]
+
+    return metadata
```

Comment:
(If continue above is removed) we get the following warnings printed out in relation to the metadata:
`WARNING: instrument key not in attributes or metadata`
`WARNING: inlet_height_magl key not in attributes or metadata`

Should "inlet_height_magl" (as the same as inlet in this case) and "instrument" (perhaps as "NOT_SET" if this is not available in the metadata) be added here for consistency?